### PR TITLE
Enable CDC for Charges db and tables. Pre- and PostDeploy scripts for DbUp

### PR DESCRIPTION
**Pull request by HenrikSommer**
## Description

Enable CDC for Charges database and tables using pre- and postdeploy-scripts for DbUp.

In a later PR it will be possible to enable pre- and postdeploy scripts with DbUp application parameters.

### CODE REVIEW

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.ApplyDBMigrationsApp/Program.cs [hunk 1]
```diff
@@ -22,15 +24,27 @@ public static class Program
...
-            var filter = EnvironmentFilter.GetFilter(args);
             var isDryRun = args.Contains("dryRun");
 
+            var preDeployResult = PerformUpgrade(connectionString, EnvironmentFilter.GetPreDeployFilter(args), isDryRun);
+            if (!preDeployResult.Successful) { return ResultReporter.ReportResult(preDeployResult); }
+            ResultReporter.ReportResult(preDeployResult);
+
+            var result = PerformUpgrade(connectionString, EnvironmentFilter.GetFilter(args), isDryRun);
+            if (!preDeployResult.Successful) { return ResultReporter.ReportResult(result); }
+            ResultReporter.ReportResult(result);
```
Comment (prtandrup) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.ApplyDBMigrationsApp/Program.cs:35]: Think this a copy paste thing

```suggestion
            var result = PerformUpgrade(connectionString, EnvironmentFilter.GetFilter(args), isDryRun);
            if (!result.Successful) { return ResultReporter.ReportResult(result); }
            ResultReporter.ReportResult(result);
```

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.ApplyDBMigrationsApp/Program.cs [hunk 2]
```diff
@@ -22,15 +24,27 @@ public static class Program
...
             var connectionString = ConnectionStringFactory.GetConnectionString(args);
-            var filter = EnvironmentFilter.GetFilter(args);
             var isDryRun = args.Contains("dryRun");
 
+            var preDeployResult = PerformUpgrade(connectionString, EnvironmentFilter.GetPreDeployFilter(args), isDryRun);
+            if (!preDeployResult.Successful) { return ResultReporter.ReportResult(preDeployResult); }
+            ResultReporter.ReportResult(preDeployResult);
+
+            var result = PerformUpgrade(connectionString, EnvironmentFilter.GetFilter(args), isDryRun);
+            if (!preDeployResult.Successful) { return ResultReporter.ReportResult(result); }
```
Comment (BjarkeMeier) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.ApplyDBMigrationsApp/Program.cs:34]: I guess that this bug(?) and the increasing complexity of the Program class suggest that we ought to have some unit tests.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.ApplyDBMigrationsApp/Program.cs [hunk 3]
```diff
@@ -22,15 +24,27 @@ public static class Program
         public static int Main(string[] args)
         {
             var connectionString = ConnectionStringFactory.GetConnectionString(args);
-            var filter = EnvironmentFilter.GetFilter(args);
             var isDryRun = args.Contains("dryRun");
 
+            var preDeployResult = PerformUpgrade(connectionString, EnvironmentFilter.GetPreDeployFilter(args), isDryRun);
+            if (!preDeployResult.Successful) { return ResultReporter.ReportResult(preDeployResult); }
+            ResultReporter.ReportResult(preDeployResult);
```
Comment (BjarkeMeier) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.ApplyDBMigrationsApp/Program.cs:31]: nit: Duplicate code - you execute ResultReporter.ReportResult() regardless of the preDeployResult status.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.ApplyDBMigrationsApp/Helpers/EnvironmentFilter.cs [hunk 4]
```diff
@@ -19,12 +19,24 @@ namespace GreenEnergyHub.Charges.ApplyDBMigrationsApp.Helpers
 {
     public static class EnvironmentFilter
     {
+        public static Func<string, bool> GetPreDeployFilter(string[] args)
```
Comment (BjarkeMeier) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.ApplyDBMigrationsApp/Helpers/EnvironmentFilter.cs:22]: @bemwen, what are the thoughts on this tool diverging in different repos?

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.ApplyDBMigrationsApp/Scripts/PostDeploy/202104131400 Enable Table CDC.sql [hunk 5]
```diff
@@ -0,0 +1,41 @@
+﻿EXEC sys.sp_cdc_enable_table  
+@source_schema = N'dbo',
```
Comment (BjarkeMeier) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.ApplyDBMigrationsApp/Scripts/PostDeploy/202104131400 Enable Table CDC.sql:2]: Perhaps this comment addresses a topic related to other work.
Are we using the dbo schema? How about using e.g. 'charges' and 'market-participants'?

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.ApplyDBMigrationsApp/Scripts/PostDeploy/202104131400 Enable Table CDC.sql [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+
+EXEC sys.sp_cdc_enable_table  
+@source_schema = N'dbo',
+@source_name   = N'VATPayerType',
+@role_name = NULL
+GO
+
+EXEC sys.sp_cdc_enable_table  
+@source_schema = N'dbo',
+@source_name   = N'MarketParticipant',
```
Comment (BjarkeMeier) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.ApplyDBMigrationsApp/Scripts/PostDeploy/202104131400 Enable Table CDC.sql:21]: nit: Do we need cdc on market participant?

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.ApplyDBMigrationsApp/Scripts/PostDeploy/202104131400 Enable Table CDC.sql [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+
+EXEC sys.sp_cdc_enable_table  
+@source_schema = N'dbo',
+@source_name   = N'ChargePrice',
+@role_name = NULL
+GO
+
+EXEC sys.sp_cdc_enable_table  
+@source_schema = N'dbo',
+@source_name   = N'ValidationRuleConfiguration',
```
Comment (BjarkeMeier) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.ApplyDBMigrationsApp/Scripts/PostDeploy/202104131400 Enable Table CDC.sql:39]: nit: Do we need cdc here?

**Review by BjarkeMeier — APPROVED** [review]
LGTM. You are free to decide what you want to do about the nits :-)

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.ApplyDBMigrationsApp/Program.cs (hunk 1 above)
Comment (HenrikSommer) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.ApplyDBMigrationsApp/Program.cs]: Correct, bug is fixed now.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.ApplyDBMigrationsApp/Scripts/PostDeploy/202104131400 Enable Table CDC.sql (hunk 6 above)
Comment (prtandrup) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.ApplyDBMigrationsApp/Scripts/PostDeploy/202104131400 Enable Table CDC.sql:21]: Good question. Well the companies do tend to change name, companies merge or go out of business. Shall we use CDC for tracking this?

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.ApplyDBMigrationsApp/Scripts/PostDeploy/202104131400 Enable Table CDC.sql (hunk 7 above)
Comment (prtandrup) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.ApplyDBMigrationsApp/Scripts/PostDeploy/202104131400 Enable Table CDC.sql:39]: I suggest having a short team talk, discussing whether to CDC it all or not :)

**Review by prtandrup — APPROVED** [review]
LGTM

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.ApplyDBMigrationsApp/Scripts/PostDeploy/202104131400 Enable Table CDC.sql (hunk 6 above)
Comment (BjarkeMeier) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.ApplyDBMigrationsApp/Scripts/PostDeploy/202104131400 Enable Table CDC.sql:21]: Actually, I'm curious about the shift from the POC story to actual implementation in production code. Did someone make a decision that charges rely on cdc for audit?

**Pull request merged**